### PR TITLE
Introduces new utility MoreFiles#listDirectoryFilesRecursive

### PR DESCRIPTION
**Pull request by jonsyu1**
This is a file utility for recursively listing the files in a directory as a
flattened list of `Path`s.

I thought about how to write unit tests for this, but this seems painful without
a mocked `Path` and `FileSystem`. Any recommendations?

### CODE REVIEW

**Comment by jonsyu1** [comment]
Tested this by hand by recursively listing my `JAVA_HOME`. Would be open to suggestions on how to contribute unit tests: wanted a mocked `Path` and `FileSystem`, but rolling my own seems like a lot of effort for one test.

**Comment by bluekeyes** [comment]
Unfortunately, there's not a great way to test things like this - you're right that mocking everything would be a lot of effort. For the most part, I assume the correctness of the local file system implementations and test things on files that either exist in a `test/resources` directory or are created as part of the test. I'll look at the rest of this in more detail next week.

**Comment by bluekeyes** [comment]
Looking at this, it seems fairly specific (while useful) and somewhat contradicts `listDirectory` by only including files. What do you think of the following?

- `listDirectory`: returns all files and directories in a directory
- `listTree`: returns all files and directories in a directory and sub-directories
- `listDirectoryFiles`: returns all files in a directory
- `listTreeFiles`: returns all files in a directory and sub-directories

The `listTree*` methods could also be named `listRecursiveDirectory` and `listRecursiveDirectoryFiles` if you think "tree" is overloaded.

If this makes sense to you, feel free to only implement the method you care about (`listTreeFiles`) and I'll implement the other ones.

File: fs-base/src/main/java/com/palantir/giraffe/file/MoreFiles.java [hunk 1]
```diff
@@ -569,6 +570,56 @@ public static boolean isEmpty(Path path) throws IOException {
...
+        public List<Path> getPaths() {
+            return paths.build();
+        }
+    }
+
+    /**
+     * Given a {@code Path} representing a directory, retrieve the list of files
+     * and recursively contained in that {@code Path} as a {@code List}. The
+     * {@code Path} objects are obtained as if by resolving the name of the
+     * directory objects against the given {@code Path}.
```
Comment (bluekeyes) [inline fs-base/src/main/java/com/palantir/giraffe/file/MoreFiles.java]: Ugh, I didn't realize the docs for `listDirectory` were this bad - I'll have to fix those. For this method, maybe something like:

"Gets a list of the files in the given directory and all sub-directories. The returned paths are relative to the given directory."

Comment (jonsyu1) [inline fs-base/src/main/java/com/palantir/giraffe/file/MoreFiles.java]: +1

**Comment by jonsyu1** [comment]
I like that separation, but I have 2 stray thoughts:
* would it make sense to combine these methods into 2 methods that take a `recursive` argument?
* if not, I'm not the biggest fan of the proposed names, because tree is overloaded, and `listRecursiveDirectory`, while fairly obvious, uses recursive as an adjective on directory, whereas it really should be an adverb on list. `recursivelyListDirectory` makes it hard to find with autocomplete and `listDirectoryRecursively` seems to be my preferred alternative at the moment.

Thoughts?

**Comment by bluekeyes** [comment]
- I definitely prefer avoiding flag arguments where possible
- While I like the shortness of "tree", I now think "recursive" is better because it's consistent with `copyRecursive`, `deleteRecursive`, and `(set|add|remove)PermissionRecursive` (even if it's not grammatical)

So the proposal is:

- `listDirectory`
- `listDirectoryRecursive`
- `listDirectoryFiles`
- `listDirectoryFilesRecursive` (long, but consistent)

**Comment by jonsyu1** [comment]
cool, amending my PR now

File: fs-base/src/main/java/com/palantir/giraffe/file/MoreFiles.java [hunk 2]
```diff
@@ -569,6 +570,55 @@ public static boolean isEmpty(Path path) throws IOException {
...
+         * The returned path list can change between calls if this visitor visits any paths.
+         */
+        public List<Path> getPaths() {
+            return paths.build();
+        }
+    }
+
+    /**
+     * Gets a list of the files in the given directory and all sub-directories.
+     * The returned paths are relative to the given directory.
```
Comment (jonsyu1) [inline fs-base/src/main/java/com/palantir/giraffe/file/MoreFiles.java]: Oops, I totally lied: I was returning absolute paths.

`[/Library/Java/JavaVirtualMachines/jdk1.7.0_71.jdk/Contents/Home/jre/bin/java, /Library/Java/JavaVirtualMachines/jdk1.7.0_71.jdk/Contents/Home/jre/bin/keytool, ...`

I'll patch this real quick.

Comment (jonsyu1) [inline fs-base/src/main/java/com/palantir/giraffe/file/MoreFiles.java]: Updated and tested this locally against `JAVA_HOME`:

`[bin/java, bin/keytool, ... ]`

**Comment by jonsyu1** [comment]
Looks like @bluekeyes gets back on Thursday (the 10th). Will check back in then for next steps.

**Pull request merged**